sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreCheckpointFormatV2Suite.scala

```diff
@@ -0,0 +1,1103 @@
+          assert(stateStoreCkptIds.drop(1).iterator
+            .sameElements(baseStateStoreCkptIds.dropRight(1)))
+      }
+
+      val versionToUniqueIdFromStateStore = Seq(1, 2).map {
+        batchVersion =>
+          val res = pickedCheckpointInfoList
+            .filter(_.batchVersion == batchVersion).map(_.stateStoreCkptId.get)
+
+          // batch Id is batchVersion - 1
+          batchVersion - 1 -> res.toArray
+      }.toMap
+
+      val commitLogPath = new Path(
+        new Path(checkpointDir.getAbsolutePath), "commits").toString
+
+      val commitLog = new CommitLog(spark, commitLogPath)
+      val metadata_ = commitLog.get(Some(0), Some(1)).map(_._2)
+
+      val versionToUniqueIdFromCommitLog = metadata_.zipWithIndex.map { case (metadata, idx) =>
+        // Use stateUniqueIds(0) because there is only one state operator
+        val res2 = metadata.stateUniqueIds(0).map { uniqueIds =>
+          uniqueIds
+        }
+        println("wei== res2")
+        res2.foreach(x => for (elem <- x) {
+          println(elem)
+        })
+        idx -> res2
+      }.toMap
+
+      versionToUniqueIdFromCommitLog.foreach {
+        case (version, uniqueIds) =>
+          versionToUniqueIdFromStateStore(version).sameElements(uniqueIds)
+      }
+    }
+  }
+
+  // This test verifies when there are task retries, the unique ids actually stored in
+  // the commit log is the same as those recorded by CkptIdCollectingStateStoreWrapper
+  testWithCheckpointInfoTracked(s"checkpointFormatVersion2 validate ID - task retry") {
+    withTempDir { checkpointDir =>
+      val inputData = MemoryStream[Int]
+      val aggregated =
+        inputData
+          .toDF()
+          .groupBy($"value")
+          .agg(count("*"))
+
+      val writer = (ds: DataFrame, batchId: Long) => {
+        val _ = ds.rdd.filter { x =>
+          val context = TaskContext.get()
+          // Retry in the first attempt
+          if (context.attemptNumber() == 0) {
+            throw new RuntimeException(s"fail the task at " +
+              s"partition ${context.partitionId()} batch Id: $batchId")
+          }
+          x.length >= 0
+        }.collect()
+      }
+
+      val query = aggregated.writeStream
+        .foreachBatch(writer)
+        .option("checkpointLocation", checkpointDir.getAbsolutePath)
+        .outputMode("update")
+        .start()
+
+      inputData.addData(1 to 100)
+      query.processAllAvailable()
+
+      inputData.addData(1 to 100)
+      query.processAllAvailable()
+
+      query.stop()
+
+      val checkpointInfoList = CkptIdCollectingStateStoreWrapper.getStateStoreCheckpointInfos
+      // scalastyle:off line.size.limit
+      // Since every task is retried once, for each partition in each batch, we should have two
+      // state store checkpointInfo. Since these infos are appended sequentially, we can group them
+      // by partitionId and batchVersion, and pick the first one as they are the retried ones.
```

Comment:
I don't quite understand why we are taking "the retried ones" here and compare with the commit log. Is it because our current implementation happens to take the first one and put in commit log?

Reply:
Yes i believe so